Mathlib/Algebra/Opposites.lean

```diff
@@ -0,0 +1,393 @@
+@[to_additive "The element of `α` represented by `x : αᵃᵒᵖ`."]
+def unop : αᵐᵒᵖ → α :=
+  id
+#align mul_opposite.unop MulOpposite.unop
+
+--attribute [pp_nodot] AddOpposite.op AddOpposite.unop
+
+@[simp, to_additive]
+theorem unop_op (x : α) : unop (op x) = x :=
+  rfl
+#align mul_opposite.unop_op MulOpposite.unop_op
+
+@[simp, to_additive]
+theorem op_unop (x : αᵐᵒᵖ) : op (unop x) = x :=
+  rfl
+#align mul_opposite.op_unop MulOpposite.op_unop
+
+@[simp, to_additive]
+theorem op_comp_unop : (op : α → αᵐᵒᵖ) ∘ unop = id :=
+  rfl
+#align mul_opposite.op_comp_unop MulOpposite.op_comp_unop
+
+@[simp, to_additive]
+theorem unop_comp_op : (unop : αᵐᵒᵖ → α) ∘ op = id :=
+  rfl
+#align mul_opposite.unop_comp_op MulOpposite.unop_comp_op
+
+/-- A recursor for `mul_opposite`. Use as `induction x using mul_opposite.rec`. -/
+@[simp, to_additive "A recursor for `add_opposite`. Use as `induction x using add_opposite.rec`."]
+protected def rec {F : ∀ _ : αᵐᵒᵖ, Sort v} (h : ∀ X, F (op X)) : ∀ X, F X := fun X => h (unop X)
+#align mul_opposite.rec MulOpposite.rec
+
+/-- The canonical bijection between `α` and `αᵐᵒᵖ`. -/
+@[to_additive "The canonical bijection between `α` and `αᵃᵒᵖ`.",--]
+  simps (config := { fullyApplied := false }) apply]
+def opEquiv : α ≃ αᵐᵒᵖ :=
+  ⟨op, unop, unop_op, op_unop⟩
+#align mul_opposite.op_equiv MulOpposite.opEquiv
+
+-- porting note: We have to provide this lemma by hand, since the autogenerated version makes the
+-- simpNF linter unhappy
+lemma opEquiv_symm_apply : (opEquiv.symm.toFun : αᵐᵒᵖ → α) = unop := rfl
+
+@[to_additive]
+theorem op_bijective : Bijective (op : α → αᵐᵒᵖ) :=
+  opEquiv.bijective
+#align mul_opposite.op_bijective MulOpposite.op_bijective
+
+@[to_additive]
+theorem unop_bijective : Bijective (unop : αᵐᵒᵖ → α) :=
+  opEquiv.symm.bijective
+#align mul_opposite.unop_bijective MulOpposite.unop_bijective
+
+@[to_additive]
+theorem op_injective : Injective (op : α → αᵐᵒᵖ) :=
+  op_bijective.injective
+#align mul_opposite.op_injective MulOpposite.op_injective
+
+@[to_additive]
+theorem op_surjective : Surjective (op : α → αᵐᵒᵖ) :=
+  op_bijective.surjective
+#align mul_opposite.op_surjective MulOpposite.op_surjective
+
+@[to_additive]
+theorem unop_injective : Injective (unop : αᵐᵒᵖ → α) :=
+  unop_bijective.injective
+#align mul_opposite.unop_injective MulOpposite.unop_injective
+
+@[to_additive]
+theorem unop_surjective : Surjective (unop : αᵐᵒᵖ → α) :=
+  unop_bijective.surjective
+#align mul_opposite.unop_surjective MulOpposite.unop_surjective
+
+@[simp, to_additive]
+theorem op_inj {x y : α} : op x = op y ↔ x = y :=
+  op_injective.eq_iff
+#align mul_opposite.op_inj MulOpposite.op_inj
+
+@[simp, to_additive]
+theorem unop_inj {x y : αᵐᵒᵖ} : unop x = unop y ↔ x = y :=
+  unop_injective.eq_iff
+#align mul_opposite.unop_inj MulOpposite.unop_inj
+
+variable (α)
+
+@[to_additive]
+instance [Nontrivial α] : Nontrivial αᵐᵒᵖ :=
+  op_injective.nontrivial
+attribute [instance] AddOpposite.instNontrivialAddOpposite
```

Comment:
Can you add a porting note (just on this one is sufficient) referencing a discussion for why we need to do it this way for now?

Reply:
Reference https://github.com/leanprover-community/mathlib4/issues/667

Reply:
This has been fixed in #678; please merge master and verify.

Reply:
Oh, I was sloppy and did the search & replace by hand. This was the very PR that caused me to write #678